### PR TITLE
ZSA note encryption in Orchard crate

### PR DESCRIPTION
**Pull request by naure**
Based on #2 for the NoteType.

Features:

- Encode note plaintext.
- Decode note plaintext.
- Compute note commitments.
- Backward-compatibility of the above.

See test vectors: https://github.com/QED-it/zcash-test-vectors/pull/1

### CODE REVIEW

File: src/note_encryption.rs [hunk 1]
```diff
@@ -149,13 +173,23 @@ impl Domain for OrchardDomain {
...
         np[12..20].copy_from_slice(&note.value().to_bytes());
         // todo: add note_type
         np[20..52].copy_from_slice(note.rseed().as_bytes());
-        np[52..].copy_from_slice(memo);
+        if is_native {
+            np[52..].copy_from_slice(memo);
+        } else {
+            let zsa_type = note.note_type().to_bytes();
+            np[52..84].copy_from_slice(&zsa_type);
+            let short_memo = &memo[0..memo.len() - 32];
```
Comment (PaulLaux) [inline src/note_encryption.rs:189]: I would replace `32` with `ZSA_TYPE_SIZE`

File: src/note_encryption.rs [hunk 2]
```diff
@@ -312,26 +357,73 @@ impl ShieldedOutput<OrchardDomain, COMPACT_NOTE_SIZE> for CompactAction {
...
+            |diversifier| {
+                assert_eq!(diversifier, &note.recipient().diversifier());
+                Some(*note.recipient().pk_d())
+            }
+        ).expect("Plaintext parsing failed");
+
+        // Check.
+        assert_eq!(parsed_note, note);
+        assert_eq!(parsed_recipient, note.recipient());
+        if note_type.is_native().into() {
```
Comment (PaulLaux) [inline src/note_encryption.rs:415]: should be `parsed_note.note_type.is_native().into()`.

**Review by PaulLaux — APPROVED** [review]
Approved with some comments/suggestions.

Also, all the following locations should be replaced with constants:
image

File: src/note_encryption.rs (hunk 2 above)
Comment (naure) [reply src/note_encryption.rs:415]: Fixed using `prop::bool::ANY`

**Pull request merged**